File: general-interview-prep/javascript/connectNodesAtSamelevel.js
/**
 * // Definition for a Node.
 * function Node(val, left, right, next) {
 *    this.val = val === undefined ? null : val;
 *    this.left = left === undefined ? null : left;
 *    this.right = right === undefined ? null : right;
 *    this.next = next === undefined ? null : next;
 * };
 */
/**
 * @param {Node} root
 * @return {Node}
 */
var connect = function(root) {
    let og = root;
    while(root && root.left) {
        next = root.left;

        while(root) {
            root.left.next = root.right
            root.right.next = root.next && root.next.left
            root = root.next
        }

        root = next
    }
    return og
};

File: general-interview-prep/golang/connectNodesAtSameLevel.go
/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Left *Node
 *     Right *Node
 *     Next *Node
 * }
 */

 func connect(root *Node) *Node {
  originalNode := root
  next := root
  for root != nil && root.Left != nil {
      next = root.Left
      
      for root != nil {
          root.Left.Next = root.Right
          if root.Next != nil {
              root.Right.Next = root.Next.Left
          }
          root = root.Next
      }
      
      root = next
  }
  
  return originalNode
}